spec/request/gi/institutions_request_spec.rb

```diff
@@ -48,6 +48,6 @@
       get '/v0/gi/institutions/autocomplete?term=%ADuniversity'
     end
 
-    expect(response).to have_http_status(:bad_request)
+    expect(response).to be_success
```

Comment:
is rails able to handle these encodings now, and if yes is the business logic that we should accept them?

Reply:
Rails 5 introduced [`ActionDispatch` code](https://github.com/rails/rails/commit/59ab2d1ee5995d9ea27ca60e92576518c1898c59#diff-a2247cc200a3f9cd3d0e5ef8d76c4df8 
) to 

>   Catch invalid UTF-8 querystring values and respond with BadRequest

>   Check querystring params for invalid UTF-8 characters, and raise an
    ActionController::BadRequest error if present. Previously these strings
    would typically trigger errors further down the stack.

We were catching encoding errors down the stack in this controller as seen above. I did not find other instances where this was an issue.

The previous commit in this PR introduces `utf8-cleaner` middleware to rack which strips invalid UTF8 sequences from the environment altogether.

Alternate ideas:
- We could change the test to expect a BadRequest to be raised, thereby keeping the expected flow but that would be a digression from the current exception raised
- We could introduce a rescue block in rack middleware to raise the expected `Common::Exceptions::InvalidFieldValue` error and ignore the utf8-cleaner gem

Certainly inverting the business logic is a bit cavalier, good catch @kreek - thoughts all?

cc: authors @patrickvinograd @saneshark

Reply:
so why is this a problem as part of gem upgrade? is Rails 5 identifying these as problematic higher up in the stack?

Reply:
042335b -> acf1810 -> 2def409

----

At 042335b, the failures seen when running [`rspec spec/request/gi/institutions_request_spec.rb`](https://github.com/department-of-veterans-affairs/vets-api/blob/master/spec/request/gi/institutions_requesxt_spec.rb)

```sh
 1) institutions responds to GET #search with bad encoding
     Failure/Error: status, headers, response = @app.call(env)

     ActionController::BadRequest:
       Invalid query parameters: Non UTF-8 value: ?illinois
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/request.rb:59:in `fetch'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/request.rb:59:in `fetch_header'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/session/abstract/id.rb:232:in `context'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/session/abstract/id.rb:226:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-attack-5.0.1/lib/rack/attack.rb:147:in `call'
     # ./lib/statsd_middleware.rb:13:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/olive_branch-2.0.0/lib/olive_branch/middleware.rb:14:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/etag.rb:25:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/conditional_get.rb:25:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/head.rb:12:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rails_semantic_logger-4.4.0/lib/rails_semantic_logger/rack/logger.rb:43:in `call_app'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rails_semantic_logger-4.4.0/lib/rails_semantic_logger/rack/logger.rb:28:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/runtime.rb:22:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/sendfile.rb:111:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-cors-0.4.1/lib/rack/cors.rb:81:in `call'
     # ./lib/http_method_not_allowed.rb:16:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/sentry-raven-2.7.4/lib/raven/integrations/rack.rb:51:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/railties-5.0.7.1/lib/rails/engine.rb:522:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-test-0.6.3/lib/rack/mock_session.rb:30:in `request'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-test-0.6.3/lib/rack/test.rb:244:in `process_request'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-test-0.6.3/lib/rack/test.rb:124:in `request'
     # ./spec/request/gi/institutions_request_spec.rb:20:in `block (3 levels) in <top (required)>'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/vcr-3.0.3/lib/vcr/util/variable_args_block_caller.rb:9:in `call_block'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/vcr-3.0.3/lib/vcr.rb:189:in `use_cassette'
     # ./spec/request/gi/institutions_request_spec.rb:19:in `block (2 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # Rack::QueryParser::InvalidParameterError:
     #   Non UTF-8 value: ?illinois
     #   /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/request.rb:59:in `fetch'

  2) institutions responds to GET #autocomplete with bad encoding
     Failure/Error: status, headers, response = @app.call(env)

     ActionController::BadRequest:
       Invalid query parameters: Non UTF-8 value: ?university
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/request.rb:59:in `fetch'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/request.rb:59:in `fetch_header'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/session/abstract/id.rb:232:in `context'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/session/abstract/id.rb:226:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-attack-5.0.1/lib/rack/attack.rb:147:in `call'
     # ./lib/statsd_middleware.rb:13:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/olive_branch-2.0.0/lib/olive_branch/middleware.rb:14:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/etag.rb:25:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/conditional_get.rb:25:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/head.rb:12:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rails_semantic_logger-4.4.0/lib/rails_semantic_logger/rack/logger.rb:43:in `call_app'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rails_semantic_logger-4.4.0/lib/rails_semantic_logger/rack/logger.rb:28:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/runtime.rb:22:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/sendfile.rb:111:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-cors-0.4.1/lib/rack/cors.rb:81:in `call'
     # ./lib/http_method_not_allowed.rb:16:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/sentry-raven-2.7.4/lib/raven/integrations/rack.rb:51:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/railties-5.0.7.1/lib/rails/engine.rb:522:in `call'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-test-0.6.3/lib/rack/mock_session.rb:30:in `request'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-test-0.6.3/lib/rack/test.rb:244:in `process_request'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-test-0.6.3/lib/rack/test.rb:124:in `request'
     # ./spec/request/gi/institutions_request_spec.rb:48:in `block (3 levels) in <top (required)>'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/vcr-3.0.3/lib/vcr/util/variable_args_block_caller.rb:9:in `call_block'
     # /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/vcr-3.0.3/lib/vcr.rb:189:in `use_cassette'
     # ./spec/request/gi/institutions_request_spec.rb:47:in `block (2 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # Rack::QueryParser::InvalidParameterError:
     #   Non UTF-8 value: ?university
     #   /home/kf/.rvm/gems/ruby-2.3.7@vetsapi/gems/rack-2.0.6/lib/rack/request.rb:59:in `fetch'
```
 
These errors are due to a change in [ActionDispatch](https://github.com/rails/rails/commit/59ab2d1ee5995d9ea27ca60e92576518c1898c59#diff-a2247cc200a3f9cd3d0e5ef8d76c4df8). 


--- 

This is mitigated in acf1810 within which I introduce the utf8-cleaner gem. This gem introduces a simple [Rack middleware](https://github.com/singlebrook/utf8-cleaner/blob/master/lib/utf8-cleaner/middleware.rb) that sanitizes all utf-8 strings. The addition of this gem results in the aforementioned tests failing in this manner:

```sh
  1) institutions responds to GET #search with bad encoding
     Failure/Error: expect(response).to have_http_status(:bad_request)
       expected the response to have status code :bad_request (400) but it was :ok (200)                                                                       
     # ./spec/request/gi/institutions_request_spec.rb:23:in `block (2 levels) in <top (required)>'                                                             

  2) institutions responds to GET #autocomplete with bad encoding
     Failure/Error: expect(response).to have_http_status(:bad_request)
       expected the response to have status code :bad_request (400) but it was :ok (200)                                                                       
     # ./spec/request/gi/institutions_request_spec.rb:51:in `block (2 levels) in <top (required)>'                                                             
```

These errors indicate the utf8-cleaner Rack middleware is successfully sanitizing invalid utf-8 strings, and when we reach the custom method in the [`GIController`](https://github.com/department-of-veterans-affairs/vets-api/blob/master/app/controllers/gi_controller.rb#L18) we no longer raise the `InvalidFieldValue` exception.

```rb
  def safe_encoded_params(input)
    input.transform_values do |v|
      begin
        v.respond_to?(:encode) && v.encode!('UTF-8', 'binary')
      rescue EncodingError
        raise Common::Exceptions::InvalidFieldValue.new('parameter', v.scrub)
      end
    end
  end    
```
--- 

The `#safe_encoded_params` method code is removed in 2def409 - which triggers a few decision points:

- do we want to accept invalid utf-8 strings in queries and sanitize them ourselves, per-controller?
- does the `GIController` have a specific, undocumented need for us to accept these invalid strings?
- are we comfortable simply changing the business logic to accept these now that utf8-cleaner will handle them?
- do we want to write our own middleware that either sanitizes or raises the `Common::Exceptions::InvalidFieldValue` exception and maintains the business logic?


Hope that helps

Reply:
I'm not sure we should be sanitizing search parameters, I suppose it's ok as long as we return the sanitized search parameters in the response and the FE can show that the search params were modified. I think in general, if you're sanitizing user input, it should be clear to the user that you sanitized inputs... and since we're sort of a Single Page App, maybe it should be up to the FE to do this, not the BE. That way the FE can keep track of that state change and let the user know that it has changed the parameters.